internal/blocktx/processor.go

```diff
@@ -525,27 +486,118 @@ func (ph *Processor) processBlock(msg *p2p.BlockMessage) {
 		return
 	}
 
 	// add the total block processing time to the stats
-	ph.logger.Info("Processed block", slog.String("hash", blockHash.String()), slog.Int("txs", len(msg.TransactionHashes)), slog.String("duration", time.Since(timeStart).String()))
+	p.logger.Info("Processed block", slog.String("hash", blockHash.String()), slog.Int("txs", len(msg.TransactionHashes)), slog.String("duration", time.Since(timeStart).String()))
+}
+
+func (p *Processor) getPrevBlock(ctx context.Context, prevHash, blockHash *chainhash.Hash) (*blocktx_api.Block, error) {
+	prevBlock, err := p.store.GetBlock(ctx, prevHash)
+
+	if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+		_, errDel := p.store.DelBlockProcessing(ctx, blockHash, p.hostname)
+		if errDel != nil {
+			p.logger.Error("failed to delete block processing - after getting previous block failed", slog.String("hash", blockHash.String()), slog.String("err", errDel.Error()))
+		}
+		return nil, err
+	}
+
+	return prevBlock, nil
+}
+
+func (p *Processor) longestTipExists(ctx context.Context) (bool, error) {
+	_, err := p.store.GetChainTip(ctx)
+	if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+		return false, err
+	}
+
+	if errors.Is(err, store.ErrBlockNotFound) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (p *Processor) competingChainsExist(ctx context.Context, block *blocktx_api.Block) (bool, error) {
+	if block.Status == blocktx_api.Status_ORPHANED {
+		return false, nil
+	}
+
+	if block.Status == blocktx_api.Status_LONGEST {
+		competingBlock, err := p.store.GetBlockByHeight(ctx, block.Height, blocktx_api.Status_LONGEST)
+		if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+			blockHash, err := chainhash.NewHash(block.Hash)
+			if err != nil {
+				p.logger.Error("failed to create blockHash", slog.Uint64("height", block.Height), slog.String("err", err.Error()))
+				return false, err
+			}
+			_, errDel := p.store.DelBlockProcessing(ctx, blockHash, p.hostname)
+			if errDel != nil {
+				p.logger.Error("failed to delete block processing - after checking for competing chains", slog.String("hash", blockHash.String()), slog.String("err", errDel.Error()))
+			}
+			return false, err
+		}
+
+		if competingBlock != nil && !bytes.Equal(competingBlock.Hash, block.Hash) {
+			return true, nil
+		}
+
+		return false, nil
+	}
+
+	// If STALE status
+	return true, nil
+}
+
+func (p *Processor) hasGreatestChainwork(ctx context.Context, incomingBlock *blocktx_api.Block) (bool, error) {
+	tip, err := p.store.GetChainTip(ctx)
+	if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+		blockHash, err := chainhash.NewHash(incomingBlock.Hash)
+		if err != nil {
+			p.logger.Error("failed to create blockHash", slog.Uint64("height", incomingBlock.Height), slog.String("err", err.Error()))
+			return false, err
+		}
+		_, errDel := p.store.DelBlockProcessing(ctx, blockHash, p.hostname)
```

Comment:
Please refactor the `processBlock` function to avoid repetition of `DelBlockProcessing` and side effects in functions like `hasGreatestChainwork`

Reply:
Very good point, done